File: aml-backend/src/main/java/dev/abarmin/aml/dashboard/model/BlockModel.java
package dev.abarmin.aml.dashboard.model;

import dev.abarmin.aml.dashboard.domain.BlockProps;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class BlockModel {
  private long blockId;
  private long pageId;
  private BlockTypeModel blockType;

  @Valid
  private BlockProps blockProps;
  private boolean canMoveUp;
  private boolean canMoveDown;
}

File: aml-frontend-private/src/model/BlockModel.ts
import type { AvatarBlockProps } from "./AvatarBlockProps";
import type { BlockTypeModel } from "./BlockTypeModel";
import type { HeaderBlockProps } from "./HeaderBlockProps";
import type { LinkButtonBlockProps } from "./LinkButtonBlockProps";
import type { SocialNetworksBlockProps } from "./SocialNetworksBlockProps";

export type AllowedBlockProps = HeaderBlockProps | LinkButtonBlockProps | AvatarBlockProps | SocialNetworksBlockProps;

export class BlockResponse {
  blockId: number;
  pageId: number;
  blockType: BlockTypeModel;
  blockProps: AllowedBlockProps;
  canMoveUp: boolean;
  canMoveDown: boolean;

  constructor(
    blockId: number,
    pageId: number,
    blockType: BlockTypeModel,
    blockProps: AllowedBlockProps,
    canMoveUp: boolean,
    canMoveDown: boolean
  ) {
    this.blockId = blockId;
    this.pageId = pageId;
    this.blockType = blockType;
    this.blockProps = blockProps;
    this.canMoveUp = canMoveUp;
    this.canMoveDown = canMoveDown;
  }
}
